### ISSUE TITLE
Enable shutdown via /-/quit

### ISSUE THREAD

**Issue by beorn7** [prometheus/pushgateway#136]
We don't need /-/reload as we don't have a config file, but to be in line with the Prometheus server, let's have /-/quit. Then we also need the equivalent of `--web.enable-lifecycle` to enable/disable the endpoint. ~~That flag could also enable/disable deletion of metrics via the web UI. Should be differently named, something about mutation via HTTP endpoints.~~

**Comment by brian-brazil** [comment]
> Should be differently named, something about mutation via HTTP endpoints.

An empty PUT has the same effect as a DELETE, so I don't think this makes a difference security wise.

There's already a section on the pushgateway on our security page (https://prometheus.io/docs/operating/security/#pushgateway), it should be updated with whatever comes from this issue.

**Comment by beorn7** [comment]
Fair point. Pushgateway has mutating endpoints by definition. It really doesn't help to just disable the DELETE method (plus, that might be part of the regular workflow).
One might argue that the Delete button could be disabled. But we already have a confirmation dialogue, thus it will hardly happen that people delete by accident. If they delete on purpose, the will always be able to do so.

**Comment by snebel29** [comment]
Just adding clarification on the necessary flag for the record, from the documentation https://prometheus.io/docs/operating/security/#prometheus

> The /-/quit endpoint is disabled by default, but can be enabled with the -web.enable-remote-shutdown flag.

**Comment by snebel29** [comment]
Acually from [this post](https://www.robustperception.io/shutting-down-prometheus) and from prometheus itself [here](https://github.com/prometheus/prometheus/blob/master/cmd/prometheus/main.go#L163-L164) seems to contradict the documentation, yes the [documentation](https://prometheus.io/docs/operating/security/#prometheus) seems to be misleading about enabling this endpoint

> The --web.enable-lifecycle flag controls HTTP reloads and shutdowns of Prometheus. This is also disabled by default. If enabled they will be accessible under the /-/reload and /-/quit paths.
>
> In Prometheus 1.x, /-/reload and using DELETE on /api/v1/series are accessible to anyone with access to the HTTP API. The /-/quit endpoint is disabled by default, but can be enabled with the -web.enable-remote-shutdown flag.

I guess once fully clarified, this could be fixed in the [doc repository](https://github.com/prometheus/docs)

**Comment by beorn7** [comment]
The first quoted paragraph is for Prometheus 2.x (the current major version), the second paragraph is for Prometheus 1.x.

We want to mirror the  Prometheus 2.x behavior here.

**Comment by snebel29** [comment]
:+1: thanks for the clarification, I hope I'm not too energetic bothering everybody too much :smile:

**Comment by snebel29** [comment]
I took a look to Prometheus implementation of the `quit` endpoint, I've identified the following steps (feedback welcomed).

- [ ] Add `web.enable-lifecycle` flag to `Enable shutdown via HTTP request.`.
- [ ] Create quitCh `chan struct{}` and pass to interruptHandler function.
- [ ] Create new quit handler that close quitCh channel sending `struct{}` zero value.
- [ ] Handle POST and PUT verbs with the quit handler for `/-/quit` path if above flag is enabled, 403 otherwise.
- [ ] Introduce quit logic into [interruptHandler function](https://github.com/prometheus/pushgateway/blob/51ae91b94abbe95a5b92f3261fca945ee93f0861/main.go#L193-L199) by adding a select to pop from notifier and the new quit channel, may be rename the function name to something like `exitHandler` instead.
- [ ] Update documentation to indicate the new endpoint exists.

I may realize something is missing while implementing
